File: src/Oving7/Oving7.java
package Oving7;

import org.omg.CORBA.INTERNAL;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.stream.Stream;

public class Oving7 {
    public static void main(String[] args) throws Exception {
        Stream<String> lineStream = Files.lines(Paths.get("C:\\L7Skandinavia-navn"));
        ArrayList<String> lines = new ArrayList<>();
        lineStream.forEach(lines::add);
        int numNames = Integer.parseInt(lines.get(0).trim());
        lines.remove(0);
        HashMap<Integer, String> names = new HashMap<>(numNames);
        lines.forEach(line -> {
            String[] parts = line.trim().split("\\s+");
            names.put(Integer.parseInt(parts[0]), parts[2]);
        });

        lineStream = Files.lines(Paths.get("C:\\L7Skandinavia"));
        lines = new ArrayList<>();
        lineStream.forEach(lines::add);
        String[] num = lines.get(0).trim().split("\\s+");
        int numNodes = Integer.parseInt(num[0]);
        int numEdges = Integer.parseInt(num[1]);
        lines.remove(0);
        Graph graph = new Graph(numNodes);
        for (int i = 0; i < graph.nodes.length; i++) {
            String name = names.containsKey(i) ? names.get(i) : "" + i;
            graph.nodes[i] = new GraphNode(name);
        }

        lines.forEach(line -> {
            String[] parts = line.trim().split("\\s+");
            int fromNode = Integer.parseInt(parts[0]);
            int toNode = Integer.parseInt(parts[1]);
            graph.nodes[fromNode].edges.add(graph.nodes[toNode]);
        });

        GraphNode node = graph.nodes[1234];
        System.out.println(node.label + " har " + node.edges.size() + " veier fra seg");
    }
}

class Graph {
    GraphNode[] nodes;

    Graph(int numNodes) {
        nodes = new GraphNode[numNodes];
    }
}

class GraphNode {
    String label;
    UnorderedLinkedList edges;

    GraphNode(String label) {
        this.label = label;
        edges = new UnorderedLinkedList();
    }

    GraphNode() {
        this("");
    }
}

class UnorderedLinkedList {

    LinkedListNode head;

    LinkedListNode add(Object value) {
        LinkedListNode newNode = new LinkedListNode(value);
        newNode.next = head;
        head = newNode;
        return newNode;
    }

    void remove(LinkedListNode node) {

    }

    int size() {
        int i = 0;
        LinkedListNode current = head;
        while (current != null) {
            i++;
            current = current.next;
        }
        return i;
    }
}

class LinkedListNode {
    LinkedListNode next;
    Object value;

    LinkedListNode(Object value) {
        this.value = value;
    }
}

File: src/Oving7/Oving7.py
import sys

from Graph import Graph, GraphNode
from Queue import Queue


def bfs(graph, from_node):
    for node in graph.nodes:
        node.dist = -1

    from_node.dist = 0
    queue = Queue()
    queue.enqueue(from_node)
    while not queue.empty():
        node = queue.dequeue()
        for neighbor in node.get_neighbors():
            if neighbor.dist < 0:
                neighbor.dist = node.dist + 1
                neighbor.predecessor = node
                queue.enqueue(neighbor)


names = {}

"""
print("Reading names file...")
with open("C:\L7Skandinavia-navn", "r", encoding="utf-8") as name_file:
    name_file.readline()  # Skip første linje med antall
    for raw_line in name_file:
        parts = raw_line.rstrip().split("\t")
        names[int(parts[0])] = parts[2][1:-1]
"""


graph = Graph()
with open("C:\L7g3", "r", encoding="utf-8") as graph_file:
    num_line = graph_file.readline().split()
    num_nodes, num_edges = int(num_line[0]), int(num_line[1])
    print("{} nodes, {} edges".format(num_nodes, num_edges))

    graph.nodes = [None] * num_nodes

    print("Generating node objects...")
    for i in range(num_nodes):
        name = names[i] if i in names else i
        node = GraphNode(name)
        graph.nodes[i] = node

    print("Reading edge entries...")
    for raw_line in graph_file:
        parts = raw_line.split()
        from_node = int(parts[0])
        to_node = int(parts[1])
        graph.nodes[from_node].edges.add(graph.nodes[to_node])

print("Graph generated")

start_node = graph.nodes[0]
print("Running BFS from node {}".format(start_node.label))
bfs(graph, start_node)

node = graph.nodes[10]
while node != start_node:
    print("{} -> ".format(node.label), end='')
    node = node.predecessor
